### PR TITLE
Added field isStandalone to GeneratorBuilder to allow generation of a…

### PR DESCRIPTION
**Pull request by samschendel**
… file that is not part of the source file.

### CODE REVIEW

**Comment by kevmoo** [comment]
Tests?

**Comment by samschendel** [comment]
Added two tests for this option just a moment ago.  Let me know what you think.

**Comment by kevmoo** [comment]
@jakemac53 thoughts on this?

**Comment by jakemac53** [comment]
How would you add imports in this scenario? Previously since it was a part you would just have access to  everything the original library has access to (and could add imports there if needed as well).

In general I am not opposed but I think we need to add some mechanism for adding imports as well.

**Comment by natebosch** [comment]
The mechanism for adding imports would be the same as the mechanism for adding code - we'd write it to the output.

We have some code built on top of this that works great with that approach.

**Comment by jakemac53** [comment]
The issue with that is the generator doesn't have to much control specifically over where in the file its output goes. You can return code with some imports and if you are the only generator or the first generator that will work. However if somebody added another generator before yours then you would get syntax errors.

That might not be a strict blocker, but it does make standalone generators a bit more fragile.

File: lib/src/builder.dart [hunk 1]
```diff
@@ -20,6 +20,10 @@ class GeneratorBuilder extends Builder {
       {this.generatedExtension: '.g.dart', this.isStandalone: false}) {
     // TODO: validate that generatedExtension starts with a `.'
     //       not null, empty, etc
+    if (this.generators.length > 1) {
```
Comment (jakemac53) [inline lib/src/builder.dart]: && isStandalone

**Comment by samschendel** [comment]
ArgumentError is now thrown when one attempts to create a GeneratorBuilder with more than one generator and has set isStandalone to true.  Added two tests to verify that this error is thrown under that circumstance but not thrown when two generators are used and isStandalone is false.

**Comment by jakemac53** [comment]
lgtm, have you ran the formatter?

**Comment by samschendel** [comment]
Did just now.

**Comment by jodinathan** [comment]
Hi,

Is there an example of creating a standalone file for a group of files?
All examples I found are one generated file per parsed file.

**Comment by natebosch** [comment]
See some discussion here:
https://github.com/dart-lang/build/issues/213

package:build, and therefore package:source_gen, have a restriction of 1 input -> 1 or more outputs whether the outputs can only change the file extension.

One workaround available today is to create a placeholder file and use that as the input, then use the glob api to find the group of files that are interesting and use them to generate the output.

**Pull request merged**